### PR TITLE
Change the payment model to use subscriptions so that there is a way to add to the subscription at a later date to reclaim the deposit amount

### PR DESCRIPTION
**Pull request by jemmyw**

### CODE REVIEW

**Comment by jemmyw** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;zZDtPMdx&#x2F;131-vols-can-be-charged-their-accountability-amount-after-the-event"> Vols can be charged their accountability amount after the event</a>

File: src/dispatch.ts [hunk 1]
```diff
@@ -115,84 +116,12 @@ function createHandler(seneca) {
...
 
     if (taskResponse.error || taskResponse.reject) {
-      log('queue error', pattern, taskResponse)
+      error('queue error', pattern, taskResponse)
     }
 
     return taskResponse
   }
 }
-
```
Comment (jemmyw) [inline src/dispatch.ts:124]: These changes are just where I've moved the test sections to be next to the function they're testing

File: src/firebase/firebase-sn.ts [hunk 2]
```diff
@@ -1,28 +1,115 @@
 import {
   byKey, firstByChildKey, byChildKey
 } from './collections'
+import {append, reduce} from 'ramda'
+import {Test} from "tape-async";
+import {test} from "../test/test";
 
 interface FirebaseValue {
-  [propName:string]:FirebaseValue | string | boolean | number
+  [propName:string]:FirebaseValue | string | boolean | number | FirebaseValue
```
Comment (TylorS) [inline src/firebase/firebase-sn.ts:9]: Silly but FirebaseValue is defined twice here

**Comment by TylorS** [comment]
This looks good to me

File: src/braintree.ts [hunk 3]
```diff
@@ -1,32 +1,64 @@
-import * as braintree from 'braintree-node'
+import {Gateway, GatewayOptions} from "../typings/braintree";
+const braintree = require('braintree')
```
Comment (jemmyw) [inline src/braintree.ts:2]: I've switched to using the official braintree library here as the promise based one was a bit pants

File: src/dispatch.ts [hunk 4]
```diff
@@ -223,6 +152,48 @@ test(__filename, 'createHandler', async function (t) {
   t.deepEquals(response, {pass: 'on'})
 })
 
+function createRecorder(ref:Firebase, tag:string) {
```
Comment (jemmyw) [inline src/dispatch.ts:155]: Again, only change here is moving test next to function

File: src/firebase/firebase-sn.ts [hunk 5]
```diff
@@ -1,28 +1,115 @@
...
-      push: function(values:FirebaseValue):Promise<KeyResponse> {
-         return seneca.act('role:Firebase,cmd:push', {model, values})
-      },
-      update: function(key:string, values:FirebaseValue):Promise<KeyResponse> {
-        return seneca.act('role:Firebase,cmd:update', {model, key, values})
-      },
-      remove: function(key:string):Promise<KeyResponse> {
-        return seneca.act('role:Firebase,cmd:remove', {model, key})
-      }
+class Updater {
```
Comment (jemmyw) [inline src/firebase/firebase-sn.ts:12]: Looking at this code I'm definitely feeling that this whole file should go away at some point, or at least no longer be a seneca service.

File: src/index.ts [hunk 6]
```diff
@@ -1,3 +1,6 @@
+try {
+  require("source-map-support").install()
```
Comment (jemmyw) [inline src/index.ts:2]: This was very useful for debugging!

File: src/pingpong/ping.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@ export function startPing(queueRef:Firebase) {
 
     return new Promise<boolean>(function(resolve, reject) {
       waits[key] = resolve
-      setTimeout(() => reject('timeout'), 5000)
+      setTimeout(() => reject('timeout'), 15000)
```
Comment (jemmyw) [inline src/pingpong/ping.ts:19]: Reduce the log noise

File: src/tasks/Engagements.ts [hunk 8]
```diff
@@ -1,19 +1,56 @@
 import * as assert from 'assert'
```
Comment (jemmyw) [inline src/tasks/Engagements.ts:1]: This file contains the meat of the work

**Pull request merged**